Run the multi-threaded executor at the end of each system task.

chescock opened the pull request:
# Objective

The multi-threaded executor currently runs in a dedicated task on a single thread.  When a system finishes running, it needs to notify that task and wait for the thread to be available and running before the executor can process the completion.  

See #8304

## Solution

Run the multi-threaded executor at the end of each system task.  This allows it to run immediately instead of needing to wait for the main thread to wake up.  Move the mutable executor state into a separate struct and wrap it in a mutex so it can be shared among the worker threads.  

While this should be faster in theory, I don't actually know how to measure the performance impact myself.

NthTensor commented:
I will profile this later today. Are you sure the formatting of your SAFETY comments is correct?

hymm commented:
ci failure on run-examples looks real.

chescock commented:
@NthTensor 
> I will profile this later today. 

Thanks, but it's not actually working yet, so no need just yet.  The CI caught a race condition I had missed.  

> Are you sure the formatting of your SAFETY comments is correct?

Nope, not at all!  

... Yup, the ones in `should_run` are definitely wrong.  I put the new code in between the old SAFETY comment and the `unsafe` function call.  I was confused because I didn't see an `unsafe` keyword there, but I guess it was all in an `unsafe` function.  

I'll fix those.  Were those the ones you meant, or are there other issues, too?

chescock commented:
Okay, I fixed the bug the CI caught, and added a unit test for it!  

("Dependencies / check-bans" seems to be failing for every PR and doesn't mention anything I changed, so I plan to ignore that failure.)

hymm commented:
Seems promising. Ran the many foxes example with `RUST_LOG=bevy_ecs::schedule::schedule=info,bevy_app=info,bevy_render=info`. This gets rid of the most of the spans smaller than the schedule level.

Main schedule is saving 200us, extract is 70us and render is around 70us. Over a 8ms frame time.

main
![image](https://github.com/bevyengine/bevy/assets/2180432/326c8613-643c-426a-a267-e1e368be8713)
extract
![image](https://github.com/bevyengine/bevy/assets/2180432/62227ddf-e540-425a-a8fc-8d84bee28235)
render
![image](https://github.com/bevyengine/bevy/assets/2180432/92ab5182-fa1f-4076-b59a-df72c0fd450c)
 
I'm going to run more of the examples add see how things look. I think running 3d_scene will be interesting to see how the overhead is reduced.

hymm commented on crates/bevy_tasks/src/single_threaded_task_pool.rs:
```diff
@@ -202,7 +202,7 @@ impl FakeTask {
 /// For more information, see [`TaskPool::scope`].
 #[derive(Debug)]
 pub struct Scope<'scope, 'env: 'scope, T> {
-    executor: &'env async_executor::LocalExecutor<'env>,
```
why did you need to change these lifetimes?

chescock replied:
The futures passed to `scope.spawn` may run the executor and spawn more futures, so they need `&scope`.  That only lives for `'scope`, which is shorter than `'env`.  

The change to this file makes the API match the one in `task_pool.rs`.  

(The difference between the API in `task_pool` and `single_threaded_task_pool` led to some fun and confusing compiler errors.  I somehow had rust-analyzer using a configuration with `task_pool` but `cargo build` using a configuration with `single_threaded_task_pool`, so it worked in the IDE but failed on the command line.  And the lifetime errors it reports don't point at code in `single_threaded_task_pool`, so it took me a while to figure out that this file even existed.)

hymm replied:
Might be worth pulling out of this pr then. This is a riskier change and if for some reason it gets reverted we should still keep the changes to this file.

james7132 reviewed (commented):
A more complete review is forthcoming, I'm particularly concerned about the soundness of the changes, so I may need some time to more thoroughly go through it.

With that said, I am seeing similar performance gains to what @hymm is seeing, so this is definitely looking pretty promising by itself.

james7132 commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -483,12 +555,14 @@ impl MultiThreadedExecutor {
             }
 
             // Evaluate the system set's conditions.
+            // SAFETY: We have exclusive access to ExecutorState, so no other thread is touching these conditions
```
This is not guaranteed given the function signature and the safety invariants, this likely needs to be propagated through the invariants of this function.

james7132 commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -218,67 +241,38 @@ impl SystemExecutor for MultiThreadedExecutor {
-                if block_on(poll_once(&mut executor)).is_none() {
-                    scope.spawn(executor);
-                }
+                let context = Context {
```
Is it possible to merge this with SyncUnsafeSchedule? Both of which live for the duration of the execution.

chescock replied:
Yes, I think so.  I mainly left them separate to reduce the churn, and because I wasn't sure what to name things.  

A related issue is that `Context` gets copied into each task.  It holds a reference to the `scope`, so we have to construct the `Context` inside the closure passed to `scope_with_executor`, which means it doesn't live long enough to get passed to spawned tasks by reference.  

If we want to minimize the size of the tasks, then it would make sense to create one struct outside of the `scope_with_executor` call, so that we're only copying two pointers in each task.  

I'll try to come up with some better names and reorganize.  

(The other option is to get rid of the wrapping struct and pass `scope` as a separate reference, but then you need explicit lifetimes everywhere.)

james7132 commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -522,132 +598,84 @@ impl MultiThreadedExecutor {
+        #[cfg(feature = "trace")]
+        let system_span = self.system_task_metadata[system_index]
+            .system_task_span
+            .clone();
```
Is this clone necessary?

chescock replied:
I think so, because that gets sent to a spawned task, and the borrow of `system_task_metadata` won't live long enough.  I didn't change that; the previous code did the `clone()` in the call to `task.instrument` and I just moved the call here.

james7132 replied:
OK sounds good to me.

james7132 commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -473,7 +545,7 @@ impl MultiThreadedExecutor {
         system_index: usize,
         _system: &BoxedSystem,
-        conditions: &mut Conditions,
+        conditions: &Conditions,
```
Is this 100% required to be read-only?

chescock replied:
No, probably not.  

The issue is that those conditions come from the schedule, which is only available during `run`, so they couldn't get put in the `ExecutorState` behind the mutex.  Since they were ultimately boxed systems, I figured I could treat them like the other systems.  But, yeah, this is hard to prove safe.  We need *something* shared because every task needs a copy to run the executor with, but it can be better than this.  

Hmm, I could create a `Mutex<&mut Conditions>` for the duration of `run` and pass a shared reference to that alongside the other mutex, then lock one after the other.  Would that be cleaner?  That would also address your concerns about the SAFETY comments.

james7132 reviewed (commented):
Generally looks good to me, though the potential for aliasing on `Conditions` is a bit concerning.

Ideally, we wouldn't yield back to the async executor or the OS, but we can leave that for another PR.

chescock commented:
I added two commits to try to address @james7132's review comments, then rebased the whole thing to resolve merge conflicts.

james7132 commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -111,12 +128,20 @@ pub struct MultiThreadedExecutor {
+// These all need to outlive 'scope in order to be sent to new tasks,
+// and keeping them all in a struct means we can use lifetime elision.
+#[derive(Copy, Clone)]
+pub struct Context<'scope, 'env: 'scope, 'sys> {
```
Does this need to be pub?

```suggestion
struct Context<'scope, 'env: 'scope, 'sys> {
```

james7132 commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -208,69 +234,34 @@ impl SystemExecutor for MultiThreadedExecutor {
-            systems,
-            mut conditions,
-        } = SyncUnsafeSchedule::new(schedule);
+        let environment = &Environment::new(self, schedule, world);
```
```suggestion
        let environment = Environment::new(self, schedule, world);
```
Is this borrow needed?

chescock replied:
`Context` needs a borrow because `Environment` owns the `Mutex<Conditions>`, and so that we only copy one pointer into each task instead of the whole environment.  And doing the borrow here avoids having to spell out `environment: &environment` when constructing the `Context`.

james7132 reviewed (commented):
This likely requires another check to ensure the performance hasn't degraded from the extra mutex, and we can come back and replace it with the old SyncUnsafeCells if we can reasonably prove out the safety invariants, but this otherwise looks good to me.

james7132 commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -279,16 +270,63 @@ impl SystemExecutor for MultiThreadedExecutor {
+        system: &BoxedSystem,
+    ) {
+        // tell the executor that the system finished
+        self.environment
```
This doesn't have to be done in this PR, but we have the SystemResult here, and we may be able to lock the executor, sending this through the completion queue when we can just pass it in via a function argument may be wasteful and may contribute to contention on the queue.

Perhaps we could only add the result to the queue if and only if we failed to get the lock on the executor state.

james7132 replied:
Nevermind, just tried this myself, seems to deadlock.

chescock replied:
Yeah, there's a race condition if one thread fails to get the lock, then the other thread exits the executor, and then the first thread pushes to the queue.

james7132 reviewed (approved):
Checked this against main, the new mutex does seem to eat into the improvements we saw earlier. Tested on many_foxes. Yellow is this PR, red is main.

main
![image](https://github.com/bevyengine/bevy/assets/3137680/634af00f-7a37-480b-8318-ae261e18178a)

render
![image](https://github.com/bevyengine/bevy/assets/3137680/b7264431-25da-4458-8bd2-6c247697dfc0)

There are still notable gains however. As @hymm noted, this most heavily impacts schedules with primarily small systems with little work to do.

First
![image](https://github.com/bevyengine/bevy/assets/3137680/071ebf63-eea4-49c6-9cab-bfca5ed81099)


PreUpdate
![image](https://github.com/bevyengine/bevy/assets/3137680/3d269e2c-521d-4b82-957d-d564df012f8c)

Last
![image](https://github.com/bevyengine/bevy/assets/3137680/b9be5695-52ac-4b5f-929e-575d1a2d3518)


What's interesting here is that there's almost 4x the number of calls into the multithreaded executor, but significantly improved the overall distribution of the time spent. On average cutting the time spent by 50%, and eliminating most of the long tail instances that take 100+us per run.
![image](https://github.com/bevyengine/bevy/assets/3137680/3e328ae0-28e3-481b-81f6-586d10a0d378)

---

Overall, LGTM. We can try to remove the mutex and see if we can make any improvements 
in a follow-up PR, but I'm interested in trying to minimize yielding to the task executor.

chescock commented:
Rebased to fix merge conflicts.

> the new mutex does seem to eat into the improvements we saw earlier.

Oh, wow, I wasn't expecting an uncontended mutex to actually cost anything!  One option to get back to one mutex with only safe code to would be to create a `Mutex<(&mut ExecutorState, Conditions<'_>)>`, but you'd have to pull the rest of `ExecutorState` into another struct to split the borrow.  (It doesn't sound like that's worth doing in this PR, though.)

james7132 commented:
> Oh, wow, I wasn't expecting an uncontended mutex to actually cost anything! 

Even if it's not under contention, there is still going to be a syscall to make sure that on the OS side there is no issue.

> It doesn't sound like that's worth doing in this PR, though.

Agreed. These are wins even with the impact from the extra mutex, and we can incrementally improve the results in a later PR.

hymm reviewed (commented):
Nice job. The changes are pretty conservative since it's mostly just yeeting most of the logic behind a mutex. I took the opportunity to do a pass over the safety comments. Some of them weren't quite correct before.

One thing to note is that I think this will deadlock when there aren't any taskpool threads now, since we're not awaiting in the ticking code anymore. Not sure it's relevant, but should maybe be added to the migration guide.

I want to do a little more pref testing with varying thread counts before approving. Should get to that this weekend.

chescock commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -208,69 +234,34 @@ impl SystemExecutor for MultiThreadedExecutor {
+        let systems = environment.systems;
+
+        let state = self.state.get_mut().unwrap();
+        if state.apply_final_deferred {
```
Sure, I'll move that.  

(I'm surprised that's the only one; I wasn't looking at whether the fields were used mutably, just whether they were read outside the lock.  I figured there would be something else immutable in there, but even `num_systems` gets changed.)

chescock commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -279,16 +270,63 @@ impl SystemExecutor for MultiThreadedExecutor {
+                *panic_payload = Some(payload);
+            }
+        }
+        self.tick_executor();
```
No, it spawns new tasks for each system and returns.

chescock commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -317,77 +364,114 @@ impl MultiThreadedExecutor {
             return;
         }
 
+        let mut conditions = context
```
That should be possible.  @james7132 said "We can try to remove the mutex and see if we can make any improvements
in a follow-up PR", so I'm not planning to change that in this PR.

chescock commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -317,77 +364,114 @@ impl MultiThreadedExecutor {
-                continue;
-            }
+            ready_systems.clear();
+            ready_systems.union_with(&self.ready_systems);
```
It's still necessary because `ready_systems` gets modified inside the loop by `skip_system_and_signal_dependents()` if a system is skipped.

chescock commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -516,132 +600,84 @@ impl MultiThreadedExecutor {
+        let system_meta = &self.system_task_metadata[system_index];
+
+        #[cfg(feature = "trace")]
+        let system_span = system_meta.system_task_span.clone();
```
Oh, I see, I made this span less useful by pulling the sending of the completion message out of it.  Sorry; I was trying to exclude `tick_executor()` and I had put those in the same function.  

The existing code was doing a non-blocking `try_send` on the channel, which calls `ConcurrentQueue::push` and then does some bookkeeping.  So the thread wouldn't have parked before and won't park now.  It looks like [there is a busy-wait in `push`](https://github.com/smol-rs/concurrent-queue/blob/93ee058b7f07b2346fdd55af6a033165239ef477/src/bounded.rs#L150), though, so that may be slow under contention.  

Do you want any changes here?  I could take the `tick_executor()` call out of `system_completed()` and then put `system_completed()` back in the tracing span.  Or I could remove the span.

hymm replied:
I think it's fine to leave as is for now. If we're not seeing much different here between the system run time we should probably remove it later.

hymm commented:
Ran this 3d scene and saw a somewhat ambiguous result for Render schedule. The other schedules seemed faster, but rendering is often the bottleneck so want to be carefull here.

![image](https://github.com/bevyengine/bevy/assets/2180432/3dd53c1e-7a82-4c0e-bf7a-642d1c7f4275)

I think the regression is just due to there no longer being the weird fast path hump, so we're probably ok here.

tested with `RUST_LOG=bevy_ecs::schedule::schedule=info,bevy_app=info,bevy_render=info cargo run --profile stress-test --example 3d_scene -F trace_tracy` which removes the system level tracing spans.

Looks like I was wrong about it deadlocking if there wasn't any compute threads. I added
```rust
        .add_plugins(DefaultPlugins.set(TaskPoolPlugin {
            task_pool_options: TaskPoolOptions {
                min_total_threads: 0,
                max_total_threads: 0,
                compute: TaskPoolThreadAssignmentPolicy {
                    min_threads: 0,
                    max_threads: 0,
                    percent: 0.0,
                },
                ..default()
            }
        }))
```
and it ran ok. I confirmed with tracy that there weren't any compute threads.

JoJoJet commented on crates/bevy_ecs/src/schedule/executor/multi_threaded.rs:
```diff
@@ -317,77 +364,114 @@ impl MultiThreadedExecutor {
+                self.ready_systems.set(system_index, false);
+
+                // SAFETY: `can_run` returned true, which means that:
+                // - It must have called `update_archetype_component_access` for each run condition.
```
`update_archetype_component` access is used to validate that the world matches, so it needs to be called before `run_unsafe` if you haven't already verified that the world is the same one used to initialize the system.
